app/uk/gov/hmrc/agentsubscriptionfrontend/controllers/CommonRouting.scala

```diff
@@ -15,35 +15,75 @@
  */
 
 package uk.gov.hmrc.agentsubscriptionfrontend.controllers
+
 import javax.inject.{Inject, Singleton}
-import play.api.Logger
+
+import com.kenshoo.play.metrics.Metrics
 import play.api.mvc._
 import play.api.mvc.Results._
-import uk.gov.hmrc.agentmtdidentifiers.model.Arn
+import uk.gov.hmrc.agentmtdidentifiers.model.{Arn, Utr}
 import uk.gov.hmrc.agentsubscriptionfrontend.config.AppConfig
-import uk.gov.hmrc.agentsubscriptionfrontend.models.MappingEligibility
+import uk.gov.hmrc.agentsubscriptionfrontend.connectors.MappingConnector
+import uk.gov.hmrc.agentsubscriptionfrontend.models.{InitialDetails, MappingEligibility}
 import uk.gov.hmrc.agentsubscriptionfrontend.models.MappingEligibility.{IsEligible, IsNotEligible, UnknownEligibility}
-import uk.gov.hmrc.agentsubscriptionfrontend.service.SessionStoreService
+import uk.gov.hmrc.agentsubscriptionfrontend.service.SubscriptionService
+import uk.gov.hmrc.agentsubscriptionfrontend.support.Monitoring
 import uk.gov.hmrc.http.HeaderCarrier
 
 import scala.concurrent.{ExecutionContext, Future}
 
 @Singleton
-class CommonRouting @Inject()(sessionStoreService: SessionStoreService, appConfig: AppConfig) {
-
-  private[controllers] def redirectUponSuccessfulSubscription(
-    arn: Arn)(implicit request: Request[AnyContent], hc: HeaderCarrier, ec: ExecutionContext): Future[Result] =
-    for (redirectLocation <- if (appConfig.autoMapAgentEnrolments) {
-                              sessionStoreService.fetchMappingEligible.map {
-                                MappingEligibility.apply(_) match {
-                                  case IsEligible    => routes.SubscriptionController.showLinkClients()
-                                  case IsNotEligible => routes.SubscriptionController.showSubscriptionComplete()
-                                  case UnknownEligibility => {
-                                    Logger.warn("chainedSessionDetails did not cache wasEligibleForMapping")
-                                    routes.SubscriptionController.showSubscriptionComplete()
-                                  }
-                                }
-                              }
-                            } else Future successful routes.SubscriptionController.showSubscriptionComplete())
-      yield Redirect(redirectLocation).withSession(request.session + ("arn" -> arn.value))
+class CommonRouting @Inject()(
+  mappingConnector: MappingConnector,
+  val metrics: Metrics,
+  subscriptionService: SubscriptionService,
+  appConfig: AppConfig)
+    extends Monitoring {
+
+  private[controllers] def completeMappingWhenAvailable(utr: Utr, completedPartialSub: Boolean = false)(
+    implicit request: Request[AnyContent],
+    hc: HeaderCarrier,
+    ec: ExecutionContext): Future[Result] = {
+
+    val doMappingAnswer: Boolean = request.session.get("performAutoMapping").isDefined || completedPartialSub
```

Comment:
I would prefer to keep the usage of session out of the method. It will improve the reusability of the method. I can see completeMappingWhenAvailable is used in redirectSubscriptionResponse which may not have a clue about that session.

Reply:
obtaining the performAutoMapping in session is the step before redirectSubscriptionResponse, and it is optional as it accepts the completedPartialSub parameter. essentially you will either be able to activate this method through session or available parameter. We need to store it in the session, as it is switching pages and its not all done in one step... Whereas when its called from the other place, it is capable of executing the function directly, hence why not re-using the session.